onnxruntime/core/graph/graph.cc

```diff
@@ -4084,10 +4084,73 @@
   return result;
 }
 
+void Graph::SetUpExternalInitializer(const Graph::OffsetAlignmentInfo& align_info,
+                                     size_t tensor_bytes_size,
+                                     int64_t& external_offset,
+                                     std::ofstream& external_stream,
+                                     const std::vector<uint8_t>& raw_data,
+                                     onnx::TensorProto& output_proto,
+                                     const std::filesystem::path& external_file_path,
+                                     const onnx::TensorProto& initializer,
+                                     bool is_prepacked) {
+  // update external_offset for alignment
+  // need to do padding before write actual tensor data as we do offset alignment at the begin of
+  // large tensors (offset need to be page aligned and alloction granularity aligned) like below:
+  // \242\2557\256\023.\031&0000000000000000\332)k+\253\246\342\246(&\006!\347\232\374\236\325\026\032+\36XXXX
+  // |<---small tensor---->|<---padding--->|<------------------large tensor----------------------------->|
+  if (align_info.align_offset && static_cast<int64_t>(tensor_bytes_size) > align_info.align_threshold) {
+    // Align to the larger of the page size or the allocation granularity
+    int64_t alignment_factor = std::max(static_cast<int64_t>(4096), align_info.allocation_granularity);
+    // Align to the next page or alloc granularity boundary
+    int64_t new_external_offset = static_cast<int64_t>(
+                                      std::floor((external_offset + alignment_factor - 1) / alignment_factor)) *
+                                  alignment_factor;
+
+    // padding tensor with zeros for alignment
+    for (int64_t index = external_offset; index != new_external_offset; ++index) {
+      external_stream << '0';
+    }
+
+    external_offset = new_external_offset;
+  }
+
+  for (size_t index = 0; index != tensor_bytes_size; ++index) {
+    external_stream << raw_data[index];
```

Comment:
>external_stream << raw_data[index]; [](http://example.com/codeflow?start=4&length=35)

Typically, this would convert things to text representation.
Would issuing a single write() call be better?